packages/models/src/store/json-storage.ts

```diff
@@ -0,0 +1,87 @@
+import { ChainStorage, StorageOnChain, StorageOffChain } from './chain-storage'
+
+type JSONStorageType = string | number | boolean | bigint
+export type JSONStorageOffChain =
+  | JSONStorageType
+  | {
+      [x: string]: JSONStorageType | JSONStorageType[] | JSONStorageOffChain
+    }
+
+const typeMarkLen = 4
+const typeMarkMap: Record<string, string> = {
+  string: '0001',
+  number: '0010',
+  boolean: '0011',
+  bigint: '0100',
+}
+
+function replacer(_: string, value: JSONStorageOffChain | JSONStorageType) {
+  const valueType = typeof value
+  if (valueType === 'object') return value
+  const mark = typeMarkMap[valueType]
+  if (mark) {
+    return `${mark}${value.toString()}`
+  }
+  throw new Error('Unexpected type')
+}
+
+function reviver(_: string, value: JSONStorageOffChain | JSONStorageType) {
```

Comment:
Please add a comment about the reserved parameter, or name it meaningfully with `_` prefix.